pkg/adapterManager/manager2.go

```diff
@@ -0,0 +1,443 @@
+type RuntimeAction struct {
+	// HandlerName to dispatch instances to.
+	HandlerName string
+	// Instances to dispatch to the handler.
+	Instances []*cpb.Instance
+}
+
+// HandlerState represents a ready to use handler.
+type HandlerState struct {
+	// ready to use handler.
+	Handler adapter.Handler
+	// Info that produced the above handler.
+	Info *adapter.BuilderInfo
+}
+
+// RuntimeState represents the current snapshot of the configuration database
+// and associated, initialized handlers.
+type RuntimeState interface {
+	// ResolveConfig resolves configuration to a list of actions.
+	ResolveConfig(bag attribute.Bag, variety adptTmpl.TemplateVariety) ([]*RuntimeAction, error)
+
+	// Handler returns a ready to use handler.
+	Handler(name string) (*HandlerState, bool)
+}
+
+// NewManager2 creates a new Manager2 dispatcher.
+func NewManager2(mapper expr.Evaluator, templateRepo template.Repository,
+	rt RuntimeState, gp *pool.GoroutinePool) Dispatcher {
+	m := &Manager2{
+		mapper:       mapper,
+		templateRepo: templateRepo,
+		gp:           gp,
+	}
+	m.SetRuntimeState(rt)
+	return m
+}
+
+// Manager2 is the 0.2 adapter manager.
+type Manager2 struct {
+	// mapper is the selector and expression evaluator.
+	// It is not directly used by Manager2.
+	mapper expr.Evaluator
+
+	// Repository of templates that are compiled in this Mixer.
+	// It provides a way to perform template operations.
+	templateRepo template.Repository
+
+	// <RuntimeState>
+	rtState atomic.Value
+
+	// gp is used to run manager code asynchronously.
+	gp *pool.GoroutinePool
+}
+
+// SetRuntimeState installs a new runtime state.
+func (m *Manager2) SetRuntimeState(state RuntimeState) {
+	m.rtState.Store(state)
+}
+
+// RuntimeState gets the current runtime state.
+func (m *Manager2) RuntimeState() RuntimeState {
+	v := m.rtState.Load()
+	if v == nil {
+		return nil
+	}
+	return v.(RuntimeState)
+}
+
+// Dispatcher executes aspects associated with individual API methods
+type Dispatcher interface {
+
+	// Preprocess dispatches to the set of aspects that will run before any
+	// other aspects in Mixer (aka: the Check, Report, Quota aspects).
+	Preprocess(ctx context.Context, requestBag attribute.Bag, responseBag *attribute.MutableBag) error
+
+	// Check dispatches to the set of aspects associated with the Check API method
+	Check(ctx context.Context, requestBag attribute.Bag) (*adapter.CheckResult, error)
+
+	// Report dispatches to the set of aspects associated with the Report API method
+	Report(ctx context.Context, requestBag attribute.Bag) error
+
+	// Quota dispatches to the set of aspects associated with the Quota API method
+	Quota(ctx context.Context, requestBag attribute.Bag,
+		qma *aspect.QuotaMethodArgs) (*adapter.QuotaResult2, error)
+}
+
+// dispatchInfo contains details needed to make a function call
+type dispatchInfo struct {
+	// generated code - processor.ProcessXXX ()
+	processor *template.Info
+	// Name of the handler being called.
```

Comment:
Not related to this PR, but I think we might want to  change the interface that adapter authors implement to 'Adapter' instead of Handler. It confuses me that whenever we refer to handler name, it is not the name of the adapter that is pass in via Adapter.info. Just to clarify this is the name of the handler configuration right ? Do you think we can name this handlerConfigName, I know it is too verbose, but may be will not confuse me later :-)